File: src/main/java/dev/psygamer/wireframe/util/IFreezable.java
package dev.psygamer.wireframe.util;

public interface IFreezable {
	
	/**
	 * Freezes the object.
	 *
	 * @version 1.0
	 * @since 1.0
	 */
	void freeze();
	
	/**
	 * @return Whether the object has been frozen or not.
	 * @version 1.0
	 * @since 1.0
	 */
	boolean isFrozen();
	
	final class ObjectFrozenException extends RuntimeException {
		
		public ObjectFrozenException(final Object frozenObject) {
			super("Unable to modify frozen object:" + frozenObject);
		}
		
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/util/IFreezable.kt
package dev.psygamer.wireframe.util

interface IFreezable {
	
	/** Makes the object immutable */
	fun freeze()
	
	val frozen: Boolean
	
	class ObjectFrozenException(frozenObject: IFreezable) : RuntimeException("Unable to modify frozen object: $frozenObject")
	
	companion object {
		
		@JvmStatic
		fun throwIfFrozen(freezable: IFreezable) {
			if (freezable.frozen)
				throw ObjectFrozenException(freezable)
		}
	}
}